src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.hpp.in

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * @ingroup 	{{target.string}}
+ * @defgroup	{{target.string}}_i2c I2C
+ */
+
+namespace xpcc
+{
+
+namespace stm32
+{
+
+/**
+ * I2cMaster implementation of I2C{{ id }} module.
+ *
+ * Interrupts must be enabled.
+ *
+ * @author		Georgi Grinshpun
+ * @author		Niklas Hauser
+ * @author		Sascha Schade (strongly-typed)
+ * @ingroup		{{target.string}}_i2c
+ */
+class I2cMaster{{ id }} : public ::xpcc::I2cMaster
+{
+public:
+	static const TypeId::I2cMaster{{ id }}Sda Sda;
+	static const TypeId::I2cMaster{{ id }}Scl Scl;
+
+public:
+	/**
+	 * Set up the I2C module for master operation.
+	 *
+	 * @param	rate
+	 *		`Standard` or `Fast`, `High` datarate is not supported
+	 */
+	template<class SystemClock, uint32_t baudrate=Baudrate::Standard,
+			uint16_t tolerance = xpcc::Tolerance::FivePercent >
+	static xpcc_always_inline void
+	initialize()
+	{
+		// calculate the expected clock ratio
+		constexpr uint8_t scalar = (baudrate <= 100000) ? 2 : ((baudrate <= 300000) ? 3 : 25);
+
+		// calculate the fractional prescaler value
+		constexpr float pre_raw  = float(SystemClock::I2c{{ id }}) / (scalar * baudrate);
+		// respect the prescaler range of 1 or 4 to 4095
+		constexpr uint32_t pre_ceil = std::ceil(pre_raw) > 4095 ? 4095 : std::ceil(pre_raw);
+		constexpr uint32_t pre_floor = std::floor(pre_raw) < ((scalar < 3) ? 4 : 1) ?
+				((scalar < 3) ? 4 : 1) : std::floor(pre_raw);
+
+		// calculate the possible baudrates above and below the requested baudrate
+		constexpr uint32_t baud_lower = SystemClock::I2c{{ id }} / ( scalar * pre_ceil  );
+		constexpr uint32_t baud_upper = SystemClock::I2c{{ id }} / ( scalar * pre_floor );
+
+		// calculate the fractional prescaler value corresponding to the baudrate exactly
+		// between the upper and lower baudrate
+		constexpr uint32_t baud_middle = (baud_upper + baud_lower) / 2;
+		// decide which prescaler value is closer to a possible baudrate
+		constexpr uint32_t pre = (baudrate < baud_middle) ? pre_ceil : pre_floor;
+
+		// check if within baudrate tolerance
+		constexpr uint32_t generated_baudrate = SystemClock::I2c{{ id }} / ( scalar * pre );
+		assertBaudrateInTolerance<
+				/* clostest available baudrate */ generated_baudrate,
+				/* desired baudrate */ baudrate,
+				tolerance >();
+
+		// the final prescaler value is augmented with the F/S and DUTY bit.
+		constexpr uint32_t prescaler = pre |
+				((scalar >= 3) ? (1 << 15) : 0) |
+				((scalar == 25) ? (1 << 14) : 0);
+
+		// peripheral frequency clock
+		constexpr uint8_t freq = SystemClock::I2c{{ id }} / 1000000;
+
+		// maximum rise time: assuming its linear:
+		// 1000ns @ 100kHz and 300ns @ 400kHz
+		//   => y = x * m + b, with m = -2.3333ns/kHz, b = 1233.3333ns
+		constexpr float max_rise_time = -2.333333f * (float(baudrate) / 1000.f) + 1233.333333f;
+		// calculate trise
+		constexpr float trise_raw = max_rise_time < 0 ? 0 : std::floor(max_rise_time / (1000.f / freq));
+		constexpr uint8_t trise = trise_raw > 62 ? 63 : (trise_raw + 1);
```

Comment:
I've looked at the documentation, and this is a bit more complicated than I can currently spend time on, so I suggest the following to not block this PR: Let's keep this configuration hardcoded for now, and replace these computations with a `static_assert` (and `// FIXME:`) for 100kHz baudrate at 48MHz clock.